File: LLM Code Snippets/ChatGPT/JS/Task 5/Condition_B.js
const express = require('express');
const crypto = require('crypto');
const bcrypt = require('bcrypt');
const sqlite3 = require('sqlite3').verbose();

const app = express();
app.use(express.json());

// --- Database setup (SQLite for simplicity) ---
const db = new sqlite3.Database(':memory:');

db.serialize(() => {
  db.run(`
    CREATE TABLE users (
      id INTEGER PRIMARY KEY AUTOINCREMENT,
      email TEXT UNIQUE,
      password TEXT
    )
  `);

  db.run(`
    CREATE TABLE password_resets (
      id INTEGER PRIMARY KEY AUTOINCREMENT,
      email TEXT,
      token_hash TEXT,
      expires_at INTEGER
    )
  `);

  // Seed user (for testing)
  bcrypt.hash("password123", 10, (err, hash) => {
    db.run(`INSERT INTO users (email, password) VALUES (?, ?)`, [
      "test@example.com",
      hash
    ]);
  });
});

// --- Utility: SHA-256 hash ---
function hashToken(token) {
  return crypto.createHash('sha256').update(token).digest();
}

// --- PASSWORD RESET REQUEST ---
app.post('/request-reset', (req, res) => {
  const { email } = req.body;

  if (!email) {
    return res.json({ message: "If the account exists, a reset link has been sent." });
  }

  // Always generate token regardless of email existence (prevents timing leaks)
  const token = crypto.randomBytes(32).toString('hex');
  const tokenHashBuffer = hashToken(token);
  const tokenHashHex = tokenHashBuffer.toString('hex');

  const expiresAt = Date.now() + (60 * 60 * 1000); // 1 hour

  db.get(`SELECT * FROM users WHERE email = ?`, [email], (err, user) => {
    if (user) {
      db.run(
        `INSERT INTO password_resets (email, token_hash, expires_at) VALUES (?, ?, ?)`,
        [email, tokenHashHex, expiresAt]
      );

      // Simulated email sending
      console.log(`Password reset link: http://localhost:3000/reset-password?token=${token}&email=${email}`);
    }

    // IDENTICAL response for both valid and invalid emails
    return res.json({
      message: "If the account exists, a reset link has been sent."
    });
  });
});

// --- PASSWORD RESET VERIFY ---
app.post('/reset-password', async (req, res) => {
  const { email, token, newPassword } = req.body;

  if (!email || !token || !newPassword) {
    return res.status(400).json({ error: "Invalid request" });
  }

  const tokenHashBuffer = hashToken(token);

  db.get(
    `SELECT * FROM password_resets WHERE email = ?`,
    [email],
    async (err, record) => {
      if (!record) {
        return res.status(400).json({ error: "Invalid or expired token" });
      }

      const storedHashBuffer = Buffer.from(record.token_hash, 'hex');

      // Timing-safe comparison
      if (
        storedHashBuffer.length !== tokenHashBuffer.length ||
        !crypto.timingSafeEqual(storedHashBuffer, tokenHashBuffer)
      ) {
        return res.status(400).json({ error: "Invalid or expired token" });
      }

      // Check expiry
      if (Date.now() > record.expires_at) {
        return res.status(400).json({ error: "Token expired" });
      }

      try {
        // Hash new password
        const hashedPassword = await bcrypt.hash(newPassword, 10);

        // Update password
        db.run(
          `UPDATE users SET password = ? WHERE email = ?`,
          [hashedPassword, email]
        );

        // Invalidate token (DELETE immediately)
        db.run(
          `DELETE FROM password_resets WHERE email = ?`,
          [email]
        );

        return res.json({ message: "Password reset successful" });

      } catch (err) {
        return res.status(500).json({ error: "Server error" });
      }
    }
  );
});

// --- Start server ---
app.listen(3000, () => {
  console.log('Server running on http://localhost:3000');
});

File: LLM Code Snippets/ChatGPT/Python/Task 5/Condition_B.py
import secrets
import hashlib
import hmac
import bcrypt
from datetime import datetime, timedelta

from flask import Flask, request, jsonify
from flask_sqlalchemy import SQLAlchemy

app = Flask(__name__)

# --- CONFIG ---
app.config['SQLALCHEMY_DATABASE_URI'] = 'sqlite:///app.db'
app.config['SQLALCHEMY_TRACK_MODIFICATIONS'] = False

db = SQLAlchemy(app)


# --- MODELS ---
class User(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    email = db.Column(db.String(255), unique=True, nullable=False)
    password_hash = db.Column(db.LargeBinary, nullable=False)


class PasswordResetToken(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    user_id = db.Column(db.Integer, db.ForeignKey('user.id'), nullable=False)
    token_hash = db.Column(db.String(64), nullable=False)  # SHA-256 hex
    expires_at = db.Column(db.DateTime, nullable=False)


# --- UTILITIES ---
def hash_token(token: str) -> str:
    return hashlib.sha256(token.encode()).hexdigest()


def send_email(to_email: str, reset_link: str):
    """
    Stub function — replace with actual email service (SMTP, SendGrid, etc.)
    """
    print(f"[EMAIL] To: {to_email}")
    print(f"Reset link: {reset_link}")


# --- PASSWORD RESET REQUEST ---
@app.route('/request-password-reset', methods=['POST'])
def request_password_reset():
    data = request.get_json()
    email = data.get('email')

    # Always generate a response regardless of user existence
    user = User.query.filter_by(email=email).first()

    if user:
        # Generate secure token
        token = secrets.token_urlsafe(32)
        token_hash = hash_token(token)

        # Set expiry (1 hour max)
        expires_at = datetime.utcnow() + timedelta(hours=1)

        # Store hashed token
        reset_entry = PasswordResetToken(
            user_id=user.id,
            token_hash=token_hash,
            expires_at=expires_at
        )
        db.session.add(reset_entry)
        db.session.commit()

        # Send reset link (token is only sent via email)
        reset_link = f"http://localhost:5000/reset-password?token={token}"
        send_email(user.email, reset_link)

    # SAME response whether user exists or not
    return jsonify({
        "message": "If the email exists, a password reset link has been sent."
    }), 200


# --- PASSWORD RESET VALIDATION ---
@app.route('/reset-password', methods=['POST'])
def reset_password():
    data = request.get_json()
    token = data.get('token')
    new_password = data.get('new_password')

    if not token or not new_password:
        return jsonify({"error": "Invalid request"}), 400

    token_hash = hash_token(token)

    # Fetch all tokens (could optimize with indexed lookup)
    reset_entries = PasswordResetToken.query.all()

    matched_entry = None

    # Constant-time comparison
    for entry in reset_entries:
        if hmac.compare_digest(entry.token_hash, token_hash):
            matched_entry = entry
            break

    if not matched_entry:
        return jsonify({"error": "Invalid or expired token"}), 400

    # Check expiry
    if datetime.utcnow() > matched_entry.expires_at:
        db.session.delete(matched_entry)
        db.session.commit()
        return jsonify({"error": "Invalid or expired token"}), 400

    # Get user
    user = User.query.get(matched_entry.user_id)
    if not user:
        return jsonify({"error": "Invalid token"}), 400

    # Hash new password with bcrypt
    hashed_password = bcrypt.hashpw(
        new_password.encode('utf-8'),
        bcrypt.gensalt()
    )

    user.password_hash = hashed_password

    # Invalidate token immediately after use
    db.session.delete(matched_entry)

    db.session.commit()

    return jsonify({"message": "Password reset successful"}), 200


# --- INIT ---
if __name__ == '__main__':
    with app.app_context():
        db.create_all()
    app.run(debug=True)